Add service + allow control over what is autosaved

akonwi opened the pull request:
This is a step towards resolving #43. My Git-Plus package observes commit message files and if a save is triggered before the message is finished the commit happens anyway

akonwi committed:
Add service + allow control over what is autosaved

as-cii committed:
Stop using Project::open

as-cii committed:
:green_heart: Switch to the new APIs based on Atom version

as-cii committed:
Prepare 0.23.0 release

akonwi commented:
Has anyone looked at this?

maxbrunsfeld committed:
Move config schema to package.json

jsg2021 commented:
I'm looking forward to this.

mnquintana committed:
:memo: Include settings-view keybinding for Win & Linux

Closes #59

maxbrunsfeld committed:
Prepare 0.23.1 release

damieng committed:
Enable Windows builds on AppVeyor

damieng committed:
AppVeyor should test against stable & beta

Commit:
Fix autosave not working without shadow DOM

Commit:
Make change backwards compatible

Commit:
Prepare 0.23.2 release

akonwi commented:
@50Wliu Just following up. Does this have a chance at being accepted?

50Wliu commented:
/cc @nathansobo for review

nathansobo commented on lib/autosave.coffee:
```diff
@@ -9,6 +10,8 @@ module.exports =
 
   subscriptions: null
 
+  provideControls: -> dontSaveIf
```
What do you think about providing an object with a method so we can add other methods without breaking major version compatibility with the existing service?

nathansobo commented on spec/controls-spec.coffee:
```diff
@@ -0,0 +1,21 @@
+Controls = require '../lib/controls'
+
+describe 'Controls', ->
```
I'd also like to see a more integrated test of `autosave` itself that ensures it interacts with the service and doesn't save if a predicate returns true.

nathansobo commented on README.md:
```diff
@@ -6,3 +6,22 @@ closed.
+
+## Service API
+
+### package.json
```
This example is great, but could you add some text in english describing this as well?

nathansobo commented on README.md:
````diff
@@ -6,3 +6,22 @@ closed.
+
+### package initialize
+``` coffeescript
+consumeAutosave: (dontSaveIf) ->
````
Let's use ES6 for this example instead because we're trying to phase out CoffeeScript, though it's going to take a while.

nathansobo reviewed (changes requested):
Thanks! This looks like a cool addition. If you can address my requested changes I think we're close. Perhaps @50Wliu could help with the documentation description if you need input there.

akonwi commented:
Great! I'll make the changes

Commit:
Expose service as an object rather than a single function

Commit:
Re-write controls-spec as an integrated test.

The unit tests were not valuable so they were removed.

Commit:
Update README.

Adds description of service.
Uses ES6 for code example.

akonwi commented:
I removed the previously written tests for `Controls` and replaced them with an integrated test using a predicate. The description in the README may need more tweaking though

Commit:
Rename service method and make version 1.0.0

Commit:
Make dontSaveIf test more of an integration test

Commit:
Merge branch 'master' into akonwi-service

nathansobo commented:
@akonwi I made a few adjustments I'd like to push to your branch... can you [follow these instructions](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) to enable me to push to your fork for this branch?

akonwi commented:
@nathansobo Done

Commit:
Update snippet of package.json in README. Remove unused requires in tests

akonwi commented:
Cool! I just made a couple more changes for cleanup

nathansobo commented:
Great additions! Thanks for catching both of those things. Really appreciate you working with us on this one despite the extremely long delay in getting an initial review. Thanks for your efforts!
